### PR TITLE
Network module connect ui

### PR DESCRIPTION
**Pull request by rvykydal**
Connects TUI and GUI to network module (DeviceConfigurations API).
Also fixes issues with virtual network devices in GUI along the way because network module has already got fixes of handling of NM added/removed device/connection signals.

### CODE REVIEW

File: pyanaconda/ui/tui/spokes/network.py [hunk 1]
```diff
@@ -176,66 +322,96 @@ def refresh(self, args=None):
...
+                else:
+                    device_type = nm_client.get_device_by_iface(iface).get_device_type()
+                    connection = get_default_connection(iface, device_type)
+                    connection_uuid = connection.get_uuid()
+                    log.debug("adding default connection %s for %s", connection_uuid, iface)
+                    persistent = False
+                    data = (iface, connection_uuid)
+                    nm_client.add_connection_async(connection, persistent, None,
+                                                   self._default_connection_added_cb, data)
+                    return
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/network.py]: What about joining these two returns into one return call.

File: pyanaconda/ui/gui/spokes/network.py [hunk 2]
```diff
@@ -1011,53 +819,58 @@ def _refresh_ap(self, dev_cfg, state=None):
             self._updating_device = False
 
     def _refresh_slaves(self, dev_cfg):
-        if dev_cfg.get_device_type() in [NM.DeviceType.BOND,
-                                         NM.DeviceType.TEAM,
-                                         NM.DeviceType.BRIDGE]:
+        if dev_cfg.device_type in [NM.DeviceType.BOND,
+                                      NM.DeviceType.TEAM,
+                                      NM.DeviceType.BRIDGE]:
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/network.py]: Please fix the indentation.

File: pyanaconda/ui/gui/spokes/network.py [hunk 3]
```diff
@@ -1011,53 +819,58 @@ def _refresh_ap(self, dev_cfg, state=None):
...
         speed = None
-        if dev_cfg.device:
+        device = self.client.get_device_by_iface(dev_cfg.device_name)
+        if device:
             if dev_type == NM.DeviceType.ETHERNET:
-                speed = dev_cfg.device.get_speed()
+                speed = device.get_speed()
             elif dev_type == NM.DeviceType.WIFI:
-                speed = dev_cfg.device.get_bitrate() / 1000
+                speed = device.get_bitrate() / 1000
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/network.py]: Could you please abstract this get speed elsewhere. For me it is easy to miss that first two if's are for how to get speed and the last one is about state.

File: pyanaconda/ui/tui/spokes/network.py [hunk 4]
```diff
@@ -176,66 +321,96 @@ def refresh(self, args=None):
...
+    # GetConnectionOnboot(uuid)
+    def _get_onboot(self, connection_uuid):
+        return get_onboot_from_ifcfg(connection_uuid)
+
+    def _set_onboot(self, connection_uuid, onboot):
+        return set_onboot_in_ifcfg(connection_uuid, onboot)
+
+    def _configure_connection(self, iface, connection_uuid):
+        connection = nm_client.get_connection_by_uuid(connection_uuid)
+        ui_data = WiredTUIConfigurationData()
```
Comment (poncovka) [inline pyanaconda/ui/tui/spokes/network.py]: It looks like the `ui_data` could be created in the `ConfigureNetworkSpoke` and applied in its apply method. You can get the necessary information for the `NetworkSpoke` directly from the spoke instance. Also, I don't like the name `ui_data`. The name suggests that the data has something to do with widgets or something more UI-related.

File: pyanaconda/ui/tui/spokes/network.py [hunk 5]
```diff
@@ -47,6 +55,145 @@
 
 __all__ = ["NetworkSpoke"]
 
+nm_client = NM.Client.new()
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: I'd think so, we do the same in gui. But maybe I should create the client in initialize() of the NetworkSpoke?

File: pyanaconda/ui/tui/spokes/network.py (hunk 4 above)
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: How about renaming ui_data to ui_input_data? Or ui_config_data?

File: pyanaconda/ui/tui/spokes/network.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: Hm, I should definitely make dirinstall or image install in mock guarded.

File: pyanaconda/ui/tui/spokes/network.py [hunk 6]
```diff
@@ -47,6 +54,145 @@
...
+
+        self.onboot = connection.get_setting_connection().get_autoconnect()
+
+    def update_connection(self, connection):
+        """Update NM RemoteConnection from the object.
+
+        :param connection: connection to be updated from the object
+        :type connection: NM.RemoteConnection
+        """
+
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: Fixed

File: pyanaconda/ui/tui/spokes/network.py [hunk 7]
```diff
@@ -131,33 +272,38 @@ def _activated_device_msg(self, devname):
...
-            for ipv6addr in ipv6config[0]:
-                addr_str, prefix, gateway_str = ipv6addr
-                # Do not display link-local addresses
-                if not addr_str.startswith("fe80:"):
-                    msg += _(" IPv6 Address: %(addr)s/%(prefix)d\n") % \
-                           {"addr": addr_str, "prefix": prefix}
-
+        device = nm_client.get_device_by_iface(devname)
+        if device:
+
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: Fixed

File: pyanaconda/ui/tui/spokes/network.py (hunk 1 above)
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: Fixed

File: pyanaconda/ui/gui/spokes/network.py (hunk 2 above)
Comment (rvykydal) [reply pyanaconda/ui/gui/spokes/network.py]: Fixed

File: pyanaconda/ui/gui/spokes/network.py (hunk 3 above)
Comment (rvykydal) [reply pyanaconda/ui/gui/spokes/network.py]: Hope it is more clear now.

File: pyanaconda/ui/tui/spokes/network.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: https://github.com/rhinstaller/anaconda/pull/1802/commits/d0d7035e0b7c04f5a70ac1a1f47243e01a51cb9e

File: pyanaconda/ui/tui/spokes/network.py [hunk 8]
```diff
@@ -252,58 +427,30 @@ def apply(self):
         # and we want to generate kickstart from device configurations
         # (persistent NM / ifcfg configuration), instead of using original kickstart.
         self._network_module.proxy.NetworkDeviceConfigurationChanged()
-        self._update_network_data()
-        log.debug("apply ksdata %s", self.data.network)
+
+        (valid, error) = network.sanityCheckHostname(self.hostname)
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: https://github.com/rhinstaller/anaconda/pull/1802/commits/141d03aa2bb08fc770665206208f291507a2a13c

File: pyanaconda/ui/tui/spokes/network.py (hunk 4 above)
Comment (rvykydal) [reply pyanaconda/ui/tui/spokes/network.py]: https://github.com/rhinstaller/anaconda/pull/1802/commits/09985ac46747baa3300a8ad27cdec231024acebc

File: pyanaconda/ui/tui/spokes/network.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/ui/tui/spokes/network.py]: Looks good to me.

File: pyanaconda/ui/tui/spokes/network.py (hunk 4 above)
Comment (poncovka) [reply pyanaconda/ui/tui/spokes/network.py]: Looks great. Thanks!

File: pyanaconda/ui/tui/spokes/network.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/ui/tui/spokes/network.py]: It is much better. Thanks!

**Comment by rvykydal** [comment]
@jkonecny12 I've addressed all the issues from the review.
For Vendy's review I've added 3 more patches.

File: pyanaconda/ui/gui/spokes/network.py (hunk 3 above)
Comment (jkonecny12) [reply pyanaconda/ui/gui/spokes/network.py]: Looks much better to me now. Thanks!

**Comment by rvykydal** [comment]
I did one minor update to one of the patches
https://github.com/rhinstaller/anaconda/pull/1802/commits/1740cc1e39e8a2e1de2fb8b69d141545dd9323ea#diff-bcdb5eebebb63c1f958b91c139ed52eaL71
and when the tests pass I'll merge the PR

**Pull request merged**